# Assignment 2

Imports and preprocessing

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

In [3]:
# Load MNIST
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

# Verify it loaded correctly
assert x_train.shape == (60000, 28, 28)
assert x_test.shape == (10000, 28, 28)
assert y_train.shape == (60000,)
assert y_test.shape == (10000,)

# Normalize to [0,1]
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# Flatten
x_train = x_train.reshape((x_train.shape[0], -1))
x_test = x_test.reshape((x_test.shape[0], -1))

x_train.shape, x_test.shape


((60000, 784), (10000, 784))

## Task 1: Compare 2-layer vs 3-layer vs 4-layer MLPs

### All models use 100 units per hidden layer with ReLU activation.

In [4]:
# Compare depth: 2-layer vs 3-layer vs 4-layer (hidden layers)
# 2-layer MLP
model_2 = keras.Sequential([
    layers.Input(shape=(784,)),
    layers.Dense(100, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(10, activation='softmax')
])
model_2.compile(
    optimizer=keras.optimizers.SGD(learning_rate=0.0001),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
model_2.fit(
    x_train, y_train,
    epochs=50,
    batch_size=32,
    validation_data=(x_test, y_test),
    verbose=0
)
train_loss_2, train_acc_2 = model_2.evaluate(x_train, y_train, verbose=0)
test_loss_2, test_acc_2 = model_2.evaluate(x_test, y_test, verbose=0)

# 3-layer MLP
model_3 = keras.Sequential([
    layers.Input(shape=(784,)),
    layers.Dense(100, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(10, activation='softmax')
])
model_3.compile(
    optimizer=keras.optimizers.SGD(learning_rate=0.0001),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
model_3.fit(
    x_train, y_train,
    epochs=50,
    batch_size=32,
    validation_data=(x_test, y_test),
    verbose=0
)
train_loss_3, train_acc_3 = model_3.evaluate(x_train, y_train, verbose=0)
test_loss_3, test_acc_3 = model_3.evaluate(x_test, y_test, verbose=0)

# 4-layer MLP
model_4 = keras.Sequential([
    layers.Input(shape=(784,)),
    layers.Dense(100, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(100, activation='relu'),
    layers.Dense(10, activation='softmax')
])
model_4.compile(
    optimizer=keras.optimizers.SGD(learning_rate=0.0001),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
model_4.fit(
    x_train, y_train,
    epochs=50,
    batch_size=32,
    validation_data=(x_test, y_test),
    verbose=0
)
train_loss_4, train_acc_4 = model_4.evaluate(x_train, y_train, verbose=0)
test_loss_4, test_acc_4 = model_4.evaluate(x_test, y_test, verbose=0)

print(f"2-layer | train acc: {train_acc_2:.4f} | test acc: {test_acc_2:.4f}")
print(f"3-layer | train acc: {train_acc_3:.4f} | test acc: {test_acc_3:.4f}")
print(f"4-layer | train acc: {train_acc_4:.4f} | test acc: {test_acc_4:.4f}")


2-layer | train acc: 0.8830 | test acc: 0.8886
3-layer | train acc: 0.8899 | test acc: 0.8961
4-layer | train acc: 0.8814 | test acc: 0.8865


### 2 vs 3 vs 4 layer MLP comparison

It seems that overall the 3-layer MLP performs the best on the MNIST dataset. The 4-layer has slightly degraded performance compared to the 3-layer, likely because this is a relatively simple task and 4-layers is trying to do too much. 


## Task 2: Compare different 2-layer MLP Variants

- Weight initializations
- Regularizations
- Optimizers

In [ ]:
# Weight initialization comparison on 2-layer MLP
init_methods = ["glorot_uniform", "he_normal"]
init_results = []

for init in init_methods:
    # 2-layer MLP
    model_weights_init = keras.Sequential([
        layers.Input(shape=(784,)),
        layers.Dense(100, activation='relu', kernel_initializer=init),
        layers.Dense(100, activation='relu', kernel_initializer=init),
        layers.Dense(10, activation='softmax', kernel_initializer=init)
    ])
    model_weights_init.compile(
        optimizer=keras.optimizers.SGD(learning_rate=0.0001),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy']
    )
    model_weights_init.fit(
        x_train, y_train,
        epochs=50,
        batch_size=32,
        validation_data=(x_test, y_test),
        verbose=0
    )
    train_loss_init, train_acc_init = model_weights_init.evaluate(x_train, y_train, verbose=0)
    test_loss_init, test_acc_init = model_weights_init.evaluate(x_test, y_test, verbose=0)
    init_results.append((init, train_acc_init, test_acc_init))

for init, train_acc, test_acc in init_results:
    print(f"{init:15s} | train acc: {train_acc:.4f} | test acc: {test_acc:.4f}")


glorot_uniform  | train acc: 0.8825 | test acc: 0.8882
he_normal       | train acc: 0.8906 | test acc: 0.8968


In [12]:
# Regularization comparison on 2-layer MLP
# Using dict for printing purposes
reg_methods = {
    "L2 (1e-4)": keras.regularizers.l2(1e-4),
    "L1 (1e-4)": keras.regularizers.l1(1e-4),
}
reg_results = []

for reg_name, reg in reg_methods.items():
    # 2-layer MLP
    model_reg = keras.Sequential([
        layers.Input(shape=(784,)),
        layers.Dense(100, activation='relu', kernel_regularizer=reg),
        layers.Dense(100, activation='relu', kernel_regularizer=reg),
        layers.Dense(10, activation='softmax', kernel_regularizer=reg)
    ])
    model_reg.compile(
        optimizer=keras.optimizers.SGD(learning_rate=0.0001),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy']
    )
    model_reg.fit(
        x_train, y_train,
        epochs=50,
        batch_size=32,
        validation_data=(x_test, y_test),
        verbose=0
    )
    train_loss_reg, train_acc_reg = model_reg.evaluate(x_train, y_train, verbose=0)
    test_loss_reg, test_acc_reg = model_reg.evaluate(x_test, y_test, verbose=0)
    reg_results.append((reg_name, train_acc_reg, test_acc_reg))

for reg_name, train_acc, test_acc in reg_results:
    print(f"{reg_name:15s} | train acc: {train_acc:.4f} | test acc: {test_acc:.4f}")

L2 (1e-4)       | train acc: 0.8865 | test acc: 0.8896
L1 (1e-4)       | train acc: 0.8859 | test acc: 0.8926


In [13]:
# Optimizer comparison on 2-layer MLP
opt_methods = {
    "SGD": keras.optimizers.SGD(learning_rate=0.0001),
    "Adam": keras.optimizers.Adam(learning_rate=0.0001),
}
opt_results = []

for opt_name, opt in opt_methods.items():
    # 2-layer MLP
    model_opt = keras.Sequential([
        layers.Input(shape=(784,)),
        layers.Dense(100, activation='relu'),
        layers.Dense(100, activation='relu'),
        layers.Dense(10, activation='softmax')
    ])
    model_opt.compile(
        optimizer=opt,
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy']
    )
    model_opt.fit(
        x_train, y_train,
        epochs=50,
        batch_size=32,
        validation_data=(x_test, y_test),
        verbose=0
    )
    train_loss_opt, train_acc_opt = model_opt.evaluate(x_train, y_train, verbose=0)
    test_loss_opt, test_acc_opt = model_opt.evaluate(x_test, y_test, verbose=0)
    opt_results.append((opt_name, train_acc_opt, test_acc_opt))

for opt_name, train_acc, test_acc in opt_results:
    print(f"{opt_name:15s} | train acc: {train_acc:.4f} | test acc: {test_acc:.4f}")

SGD             | train acc: 0.8821 | test acc: 0.8875
Adam            | train acc: 0.9996 | test acc: 0.9780


### Final summary

- Weight initializations
    - glorot_uniform  | train acc: 0.8825 | test acc: 0.8882
    - he_normal       | train acc: 0.8906 | test acc: 0.8968
- Regularizations
    - L2 (1e-4)       | train acc: 0.8865 | test acc: 0.8896
    - L1 (1e-4)       | train acc: 0.8859 | test acc: 0.8926
- Optimizers
    - SGD             | train acc: 0.8821 | test acc: 0.8875
    - Adam            | train acc: 0.9996 | test acc: 0.9780

The only real difference between these is the Adam optimizer. Adam achieved a near-perfect training accuracy and substantially higher test accuracy compared to SGD showing faster convergence and overall better optimization for this task.